File: python/MPPI_quadrotor.py
import numpy as np
from quadrotor import quadrotor

# ── Dynamics handle ──
DYNAMICS = quadrotor

nX = 12  # number of states
nU = 4   # number of inputs

# Quadrotor w² to force/torque matrix
kf = 8.55e-6 * 91.61
L = 0.17
b = 1.6e-2 * 91.61
m = 0.716
g = 9.81

A = np.array([
    [kf,     kf,   kf,    kf],
    [0,    L*kf,    0, -L*kf],
    [-L*kf,  0,  L*kf,     0],
    [b,     -b,    b,     -b]
])

# ── Initial and desired states ──
x0 = np.zeros(nX)
xd = np.zeros(nX)
xd[0], xd[1], xd[2] = 10.0, 10.0, 10.0

# ── MPPI parameters ──
num_samples = 1000
N = 150

utraj = np.zeros((nU, N - 1))
uOpt_list = []
xf_list = []
dt = 0.02
lam = 10.0        # lambda
nu = 1000.0
covu = np.diag([2.5, 5e-3, 5e-3, 5e-3])

R = lam * np.linalg.inv(covu)

x = x0.copy()


# ── Helper functions ──
def running_cost(x, xd, R, u, du, nu):
    Q = np.diag([2.5, 2.5, 20.0, 1.0, 1.0, 15.0, 0, 0, 0, 0, 0, 0])
    qx = (x - xd) @ Q @ (x - xd)
    J = qx + 0.5 * u @ R @ u + (1 - 1 / nu) / 2 * du @ R @ du + u @ R @ du
    return J


def final_cost(xT, xd):
    Qf = 20.0 * np.diag([2.5, 2.5, 20.0, 1.0, 1.0, 15.0, 0, 0, 0, 0, 0, 0])
    return (xT - xd) @ Qf @ (xT - xd)


# ── Run MPPI Optimization ──
for iteration in range(500):
    xf_list.append(x.copy())
    Straj = np.zeros(num_samples)
    dU = []

    # Rollouts
    for k in range(num_samples):
        du = covu @ np.random.randn(nU, N - 1)
        dU.append(du)
        xtraj = np.zeros((nX, N))
        xtraj[:, 0] = x

        for t in range(N - 1):
            u_t = utraj[:, t]
            xtraj[:, t + 1] = xtraj[:, t] + DYNAMICS(xtraj[:, t], u_t + du[:, t]) * dt
            Straj[k] += running_cost(xtraj[:, t], xd, R, u_t, du[:, t], nu)

        Straj[k] += final_cost(xtraj[:, N - 1], xd)

    minS = np.min(Straj)
    print(f"Iter {iteration:3d} | minS = {minS:.4f}", end="  ")

    # Update nominal inputs
    weights = np.exp(-1.0 / lam * (Straj - minS))
    for t in range(N - 1):
        # Gather perturbations for this timestep across all samples
        du_all = np.array([dU[k][:, t] for k in range(num_samples)])  # (num_samples, nU)
        utraj[:, t] += (weights @ du_all) / np.sum(weights)

    # Execute first control
    x = x + DYNAMICS(x, utraj[:, 0]) * dt
    uOpt_list.append(utraj[:, 0].copy())

    # Shift nominal inputs
    utraj[:, :-1] = utraj[:, 1:]
    utraj[:, -1] = 0.0

    distance = np.linalg.norm(x[:3] - xd[:3])
    print(f"| distance = {distance:.4f}")

# Convert logged data to arrays
xf = np.array(xf_list).T   # shape (nX, num_iters)
uOpt = np.array(uOpt_list).T  # shape (nU, num_iters)

File: matlab/MPPI_quadrotor.m
DYNAMICS=@quadrotor;

nX = 12;%number of states
nU = 4;%number of inputs
 
% quadrotor w^2 to force/torque matrix
kf = 8.55*(1e-6)*91.61;
L = 0.17;
b = 1.6*(1e-2)*91.61;
m = 0.716;
g = 9.81;

A = [kf, kf, kf, kf; ...
    0, L*kf, 0, -L*kf; ...
    -L*kf, 0, L*kf, 0; ...
    b, -b, b, -b];

%initial conditions
x0= [0;0;0;0;0;0;0;0;0;0;0;0];
xd= [10;10;10;0;0;0;0;0;0;0;0;0];

% Initialization
num_samples = 1000;
N = 150;

utraj = zeros(nU, N-1);
uOpt = [];
xf = [];
dt = 0.02;
lambda = 10;
nu = 1000;
covu = diag([2.5,5*1e-3,5*1e-3,5*1e-3]);

xtraj = zeros(nX, N);
R = lambda*inv(covu);

x = x0;

%% Run MPPI Optimization
for iter = 1:500
    xf = [xf,x]; % Append the simulated trajectory
    Straj = zeros(1,num_samples); % Initialize cost of rollouts
    
    % Start the rollouts and compute rollout costs
    for k = 1:num_samples
        du = covu*randn(nU, N-1);
        dU{k} = du;
        xtraj = [];
        xtraj(:,1) = x;
        for t = 1:N-1
            u = utraj(:,t);
            xtraj(:,t+1) = xtraj(:,t) + DYNAMICS(xtraj(:,t), u+du(:,t))*dt;
            Straj(k) = Straj(k) + runningCost(xtraj(:,t), xd, R, u, du(:,t), nu);
        end
        Straj(k) = Straj(k) + finalCost(xtraj(:,N), xd);
    end
    
    minS = min(Straj) % Minimum rollout cost
    
    % Update the nominal inputs
    for t = 1:N-1
        ss = 0;
        su = 0;
        for k = 1:num_samples
            ss = ss + exp(-1/lambda*(Straj(k)-minS));
            su = su + exp(-1/lambda*(Straj(k)-minS))*dU{k}(:,t);
        end
        
        utraj(:,t) = utraj(:,t) + su/ss;
    end
    
    % Execute the utraj(0)
    x = x + DYNAMICS(x, utraj(:,1))*dt;
    uOpt = [uOpt, utraj(:,1)];
    
    % Shift the nominal inputs 
    for t = 2:N-1
        utraj(:,t-1) = utraj(:,t);
    end
    utraj(:,N-1) = [0;0;0;0];
    
    distance = norm(x(1:3)-xd(1:3)) %Current distance to target
end

%% Helper functions
function J = runningCost(x, xd, R, u, du, nu)
    Q = diag([2.5, 2.5, 20, 1, 1, 15, zeros(1, 6)]);
    qx = (x-xd)'*Q*(x-xd);
    J = qx + 1/2*u'*R*u + (1-1/nu)/2*du'*R*du + u'*R*du;
end

function J = finalCost(xT,xd)
    Qf = 20*diag([2.5, 2.5, 20, 1, 1, 15, zeros(1, 6)]);
    J = (xT-xd)'*Qf*(xT-xd);
end

function plot_traj(xtraj, noisyTraj)
    % Extract positions
    pos = xtraj(1:3, :);
    pos_noisy = noisyTraj(1:3, :);

    figure;
    plot3(pos(1, :), pos(2, :), pos(3, :), 'b-', 'LineWidth', 2); hold on;
    plot3(pos_noisy(1, :), pos_noisy(2, :), pos_noisy(3, :), 'r--', 'LineWidth', 2);
    grid on; axis equal;
    xlabel('X [m]'); ylabel('Y [m]'); zlabel('Z [m]');
    legend('Original Trajectory', 'Noisy Trajectory');
    title('3D Position Trajectory of Quadrotor');
end
